### DISCUSSION TITLE
Slimproto based players (aka squeezebox)

### DISCUSSION THREAD

**Discussion by marcelveldt** [music-assistant/support#1123] [Player Support]
Documentation https://music-assistant.io/player-support/slimproto/

**Comment by disconn3ct** [comment]
This makes it impossible to use slimproto devices with the existing integration by forcibly binding the port. Is there a way to make the add-on version expose the devices for use with the rest of the system? Or disable this player so that HA continues functioning?

**Reply by heinzma** [reply]
> Is there a way to make the add-on version expose the devices for use with the rest of the system?

This is already planned: https://github.com/orgs/music-assistant/projects/2/views/1?pane=issue&itemId=22557789

**Reply by disconn3ct** [reply]
What about being able to control/disable the ports?

**Reply by marcelveldt** [reply]
There's no way to do that because 99% of the slimproto based players expect the port to be on 3483.
So just choose between MA or something else please or run both on different machines.

**Reply by disconn3ct** [reply]
100% support direct ip/port. It is rather presumptuous to assume everyone has extra computers laying around just because you do. Send me one and I'll use it.

In the meantime, this add on has fewer features than either alternative (HA or LMS) and yet it forcibly takes over a port for a service that the user might not even want. Contrast the HA slim integration, which both has configurable ports AND is not forced onto every user.

**Reply by OzGav** [reply]
You don’t have to install MA

**Reply by marcelveldt** [reply]
@disconn3ct you are aware this is an EARLY BETA ?! That said, it is not possible to do what you want. Slimproto based players can only bind to one server at the same time and the control port 3483 can not be on a different port, it is simply not supported. So even if I wanted to honor your "request", it will make no sense. MA should start fine (but without a warning) if the slimproto port is already taken.

**Reply by MarvinSchenkel** [reply]
@disconn3ct You know you can also be creative and use the 'configureable ports' feature in the HA integration to work around this 'issue'?

**Comment by disconn3ct** [comment]


**Reply by MarvinSchenkel** [reply]
I was talking about the 'Ha slim integration' (not the MA HA integration) because according to you, it has the option to change ports.

For MA it's pretty simple, we have a HUUUGE backlog of new features (like support for new music providers and players) and bugs that have a higher priority than supporting things like configurable ports.

**Reply by disconn3ct** [reply]
The HA slim integration was abandoned rather vigorously with the most recent HA release. Are you suggesting it will be fixed? If not, it has nothing to do with this bug.

Edit to clarify: When I said "HA slim integration" I meant upstream's, not the original MA. I think I got them mixed up here; upstream is not abandoned, just the "stable" MA.

**Reply by MarvinSchenkel** [reply]
I am just going by the info you provided, so no, I did not know the HA slim integration was abandoned. I am not suggesting it will be fixed. 
Also, I do not see this as 'a bug', but rather as a low-priority feature that will probably be added... someday.

**Reply by marcelveldt** [reply]
Why not just remove MA from your system, thats easy enough right ? Nobody is telling you you must use this early beta software. Just remove it and keep using the builtin slimproto integration in HA. I just have not yet finished the configuration options to disable slimproto within MA so its enabled by default. 

And I am also not understanding what you're saying by things being abandoned, that is simply not true.
You can either use HA with slimproto integration for basic player support or you can use MA for full features.

The stable version of MA broke unfortunately due to a dependency conflict so if you want to use the stable version you need to stick with HA 2023.2.x and not upgrade to 2023.2. If you want to upgrade to HA 2023.3, you have to live without MA for a while or live with the "buggy" early beta. Its a choice, your own choice.

**Reply by disconn3ct** [reply]
> I just have not yet finished the configuration options to disable slimproto

Great. That is what I originally asked: "What about being able to control/disable the ports?"

**Comment by userMak** [comment]
after installing the beta version, I see that I have to disable my slimproto / squeezebox device in order to work with MA. I am not sure that I want to. Do you believe that in the future MA could cooperate with slimproto integration?

**Reply by OzGav** [reply]
Why do you need both when MA can play to these devices? Note that the HA integration with MA will be written in the next few weeks.

**Reply by userMak** [reply]
> Why do you need both when MA can play to these devices? Note that the HA integration with MA will be written in the next few weeks.

I am using the same player for local announcements (instead of google home) so I need this as a separate entity. 

*Probably you know it but right now in my setup the although MA found all my local files (music) nothing is available to play. Only the radio works as before. ( I don't use spotify or youtube)

I have the following folder and available are only the files in the mp3 folder.

![image](https://user-images.githubusercontent.com/55662933/229425429-8ef4f2f2-c32b-474b-8288-790f8b8ed2ab.png)

All other folders although are available in browse filesystem the folders aren't opening 

![image](https://user-images.githubusercontent.com/55662933/229425811-b2b7386d-cd67-44d3-963c-52041c3f39e9.png)

**Reply by OzGav** [reply]
when the HA integration is available then you will be able to play announcements from HA via the MA players. Thus you will not need the slimproto integration when that is in place. In the meantime you will need to either roll right back to use MA1 OR do without your local announcements OR do without MA.

Please open a separate issue for the mp3 playback.

**Comment by Sirs0ri** [comment]
Because I've been messing around with getting a SqueezeLite player to run on Windows (for Music Assistant playback on that Windows PC), I wanted to document what worked for me in the end:

I am using the basic command line interface [squeezelite](https://github.com/ralph-irving/squeezelite) (builds for Windows are [here](https://sourceforge.net/projects/lmsclients/files/squeezelite/windows/)) to run a receiver on my PC with the arguments `-o 7 -M SqueezeLite -s 192.168.123.456`.

Download the binary from above and run it in a command line, no installation etc necessary. You'll need the correct IP for your Music Assistant instance, and you'll probably have to choose another output (`-o` argument) - use the `-l` argument to get a list of all audio devices, then use the number of the right one. A list of all arguments is available through `-?`, there's a lot of options to tweak the playback that I didn't look into.

This should be auto-discovered by Music Assistant and can be controlled through the Music Assistant web player. AFAIK there's no way to use local media keys to control the playback, but we automate that through home assistant anyways, right? :'D


What _didn't work_ for me was the app "Squeezelite-X" from the Microsoft store. It tries using some advanced features of the telnet interface, that aren't supported by Music Assistant, it never establishes a proper connection. There's some more info on [Discord](https://discord.com/channels/753947050995089438/1084037533781991506/1084046426125914232) incase anyone wants to read up on it.

There's a native Music Assistant app in the making (see [here](https://github.com/music-assistant/music-assistant-desktop)) but that's still pretty unstable right now, and I didn't spend a lot of time with it - it worked, until it didn't. It might be a lot more stable soon.

**Reply by Slim1399** [reply]
I hope they fix this in the next version !

**Comment by ostracizado** [comment]
Not sure if this was already talked about, but... just started using a Muse Luxe (squeezelite), without using slimproto.

It's working great, using MA. But I'm wondering about the media_player entity - there's none.
Somehow I thought that MA would create such entity so that HA could also interact with that device.

Can someone clarify that for me, please?

**Reply by OzGav** [reply]
Should have been created. Look for an entity with a _2 on the end of it (assuming there is already a HA entity of the same name)

**Reply by ostracizado** [reply]
```
023-10-26 23:31:25.914 DEBUG (MainThread) [music_assistant.providers.slimproto] Socket client connected: ('127.0.0.1', 48222)
2023-10-26 23:31:25.916 DEBUG (MainThread) [music_assistant.providers.slimproto] Player squeezeplay: XX:XX:XX:XX:XX:XX connected
2023-10-26 23:31:25.917 INFO (MainThread) [music_assistant.players] Player registered: XX:XX:XX:XX:XX:XX/squeezelite
```

I even did a new installation with squeezelite, as I was using an old firmware. And gave it a simple name's device.
After some restarts and reboots, still no new entity in the media_players.

**Reply by OzGav** [reply]
Please confirm you have the latest version of the MA integration installed from HACS and it is showing on the HA integrations page.

**Reply by ostracizado** [reply]
Ugh.... _Mea culpa_.
Just noticed that I was still using the addon BETA 2, without the custom integration.

Solved.

**Comment by peterleerental** [comment]
All our slimproto speakers have suddenly disappeared. I've reinstalled MASS and enabled Slimproto as a provider, but none of the speakers show up. Only a Chromecast dongle in the TV appears. The speakers are powered by a Raspberry Pi using Max2Play and Squeezelite and a HiFiBerry audio hat. I do not have Logitech Media Server or the Slimproto integration installed. We haven't fiddled with anything in the last few days so has there been a breaking update or something?

**Reply by OzGav** [reply]
No there hasn’t been a change with MA that would affect this to my knowledge and you are the only one reporting this. Sounds like a networking problem.

**Reply by peterleerental** [reply]
Thanks. It turned out to be that Squeezebox Server had been inadvertently turned on in the Max2Play settings on one of the players. All resolved.

**Comment by grigi** [comment]
I have a piCorePlayer setup, and added a rotary encoder so I can manually up/down volume and do things. I followed their example using `sbpd` and it seems to issue LMS commands to the LMS server, but nothing is happening.
Does Latest Beta MA (2.0.0b78) support these functions?

I'm getting this in the logs on the piCorePlayer:
```
1703359579.1898 6 discovery.c,238: Found server D200A8C0. A new address
1703359579.2905 6 discovery.c,349: discovery packet: port: 8096, name: Music Assistant, uuid: 3854c5f25d4348f6896c8c6feb9c882e
1703359579.2907 6 discovery.c,130: Squeezebox control port found: 8096
1703359579.2908 6 discovery.c,149: Server address found: 192.168.0.210
1703359618.3465 6 GPIO.c,107: Short PRESS: 138
1703359618.3466 6 control.c,134: Button CB set for button #:0, gpio pin 27
1703359618.3882 6 control.c,275: Button pressed: Pin: 27, Press Type:Short
1703359618.3883 6 servercomm.c,74: Send Command:0, Fragment:["pause"]
1703359628.6112 6 control.c,461: Encoder on GPIO 23, 24 - value: -72, detents: -36, change: 2
1703359628.6114 6 servercomm.c,74: Send Command:0, Fragment:["mixer","volume","+2"]
```
Which indicates it's talking to MA, but the commands are not doing anything that I can see?

**Reply by grigi** [reply]
I have resolved it, the example provided by piCorePlayer https://docs.picoreplayer.org/projects/control-jivelite-by-rotary-encoders-and-buttons/ is over-complicated.

All I needed to do was to use the default commands.
Basically just:
```
sbpd b,27,PLAY,2,0,NEXT,500 e,23,24,VOLU,2
```
Without `-f` as the overrided example commands don't work.

If anyone is interested I can post a working example script here.

**Reply by OzGav** [reply]
Please do. Might help people in the future.

**Reply by grigi** [reply]
To add a rotaty encoder with push button to a piCorePlayer that will play/pause, adjust volume & skip song (on long hold), wire it as: https://docs.picoreplayer.org/projects/control-jivelite-by-rotary-encoders-and-buttons/#preparation

I used GPIOs 23,24 & 27 as per DigiAMP+ documentation: https://cdn.shopify.com/s/files/1/0176/3274/files/20201209_IQaudIO_v32.pdf?v=1608196535 (page 21-23)

Install the `pcp-sbpd.tcz` & `pigpio.tcz` extensions.
Create a script in `/home/tc` called `sbpd-script.sh` on your piCorePlayer device with the following content:
```
#!/bin/sh

# start pigpiod daemon
pigpiod -t 0 -f -l -s 10

# wait for pigpiod to initialize - indicated by 'pigs t' exit code of zero

count=10 # approx time limit in seconds
while ! pigs t >/dev/null 2>&1 ; do
        if [ $((count--)) -le 0 ]; then
                printf "\npigpiod failed to initialize within time limit\n"
                exit 1 
        fi
#       printf "\nWaiting for pigpiod to initialize\n"
        sleep 1
done
printf "\npigpiod is running\n"

# load uinput module - required to be able to send keystrokes
# then set the permission to group writable, so you don't need to run sbpd with root permissions
sudo modprobe uinput
sudo chmod g+w /dev/uinput

# These GPIO pins are wired as specified in the DigiAMP+ documentation.

# button SW1
SW1=27                                                  # GPIO pin number
SH1=PLAY                                                # command for SHORT press
LO1=NEXT                                                # command for LONG press
LMS1=500                                                # milliseconds for long press

# rotary
ROT1=23
ROT2=24
MODE=2 # Detent mode - Assumes 1 dial click is x steps. 1=Step Mode (default)

# Please don't override commands using -f as Music Assistant supports a subset of the commands.
sbpd b,$SW1,$SH1,2,0,$LO1,$LMS1 e,$ROT1,$ROT2,VOLU,$MODE
```

Ensure it is executable by running `sudo chmod +x /home/tc/sbpd-script.sh`

Add the script to the user command (Bottom of `Tweaks` page, e.g. `User command #1`) and click `Save`.

Reboot, and it should work as expected.

**Comment by frelev** [comment]
I must be missing something trivial, I have LMS running as an HASS addon, players appears fine in HASS.
I assume I should have Slimproto off since I already run LMS? But how do I get the players to appear in MASS? (they do not)

**Reply by OzGav** [reply]
You cannot have LMS running at the same time for a start

**Reply by frelev** [reply]
That I understand, so question is maybe more MASS related than slimproto. How do I get my slim-devices to show up in MASS if I run LMS for other purposes.

**Reply by OzGav** [reply]
You cannot. You will be having port conflicts as MA is emulating LMS. This may change in the future.

**Comment by Dave-Casey** [comment]
### Squeezebox looping last few seconds of song
I'm using a couple of slimdevices squeezebox v3 and have found that towards the end of song playback, they get stuck repeating the last 3-5 seconds of the song, never advancing to the next track until I manual advance.  Has anyone encountered this?

**Reply by OzGav** [reply]
Have you got QUEUE FLOW MODE turned on in the player settings?

**Comment by rfporter** [comment]
I'm using MA server v2.3.0b29
I've settled on using squeezebox players because it seems to be the preferred protocol (MA client uses it).

I have noticed with SB Player (android) clients songs sometimes get stuck in a 4-5 second loop at the start of the playback. I need to manually change songs to get playback to resume. This has happened with flac and mp3 format. 

I have also noticed that grouping of squeezebox players doesn't work. It just gets stuck in the 3-4 second loop.

I have also noticed that when I try to stop playback on a SB Player, either by pausing or turning off (via MA) playback resumes on its own after a few seconds. The most reliable method to stop playback is to empty the playlist before turning off the player (via MA).

Anyone else experiencing these issues?

**Reply by rfporter** [reply]
An update: squeezebox players work when grouped under a universal group player. Of course playback is not in sync though.

When I make a standard group of 2 squeezebox players playback doesn't even start and the player status remains idle.

Here is the debug log:
2024-10-10 12:57:47.175 DEBUG (MainThread) [music_assistant.players] Handling command cmd_sync for player squeezeplay: 1c:1b:0d:7f:68:d6
2024-10-10 12:57:47.238 DEBUG (MainThread) [music_assistant.slimproto] Start serving multi-client flow audio stream for player Ryan Phone to squeezeplay: 1c:1b:0d:7f:68:d6
2024-10-10 12:57:47.761 DEBUG (MainThread) [music_assistant.slimproto] Start serving multi-client flow audio stream for player Ryan Phone to Ryan Phone
2024-10-10 12:57:47.785 DEBUG (MainThread) [music_assistant.server.providers.slimproto.multi_client_stream] Starting multi-client stream with 2/2 clients
2024-10-10 12:57:47.786 INFO (MainThread) [music_assistant.streams] Start Queue Flow stream for Queue Group1 - crossfade: False
2024-10-10 12:57:47.787 DEBUG (MainThread) [music_assistant.streams] Start Streaming queue track: jellyfin--SyLU7tNK://track/19c4f8c6cbb852a327e50e154aeeac1c (Track1) for queue Group1
2024-10-10 12:57:47.787 DEBUG (MainThread) [music_assistant.audio.media_stream] start media stream for: jellyfin--SyLU7tNK://track/19c4f8c6cbb852a327e50e154aeeac1c
2024-10-10 12:57:58.869 DEBUG (MainThread) [music_assistant.musicbrainz] get_data was delayed for 29.707 secs due to throttling
2024-10-10 12:58:09.846 DEBUG (MainThread) [aiohttp.server] Ignored premature client disconnection.
**2024-10-10 12:58:19.057 ERROR (MainThread) [ffmpeg.90] Stream error:**

**Reply by rfporter** [reply]
Another update: I have switched over to snapcast. Grouping works well. Sync is very good. Works on Windows with Snap.Net and works on android / fireTV with snapcast app.

**Reply by RandomLegend** [reply]
Hello,

i run a Music Assistant and bought 2 Squeebox Radios and one Boom - i can also connect them to my MA but can't group them. You said using snapcast works well? How did you connect your squeezebox players into snapcast?

**Reply by axodentally** [reply]
Here is the appropriate github issue: https://github.com/music-assistant/support/issues/3100
seems to be known, but the cause is not. However many changes are currently happening (see https://github.com/music-assistant/server/pull/2338 for example), so testing of beta versions or current release is probably helpful for the developers.

**Comment by Slim1399** [comment]
Hi,
I use MA Server in a Docker, I have problem for connecting "Squeezelite x" the player lists two servers, a Music Assistant server and a Home Assistant server, yet I am only one !

![Capture d'écran 2025-01-21 213425](https://github.com/user-attachments/assets/5732e1da-3911-4ddc-a263-44826d040deb)

**Reply by OzGav** [reply]
As far as I know squeezelite-x doesn’t work

**Reply by Slim1399** [reply]
just for now, one of the next updates the problem would be dealt with i hope !

**Reply by OzGav** [reply]
No you need to use the simple client.

**Reply by Slim1399** [reply]
Is there a way to have an external LMS server integrated into MA with all its Players ?

**Reply by OzGav** [reply]
No. Any basic slimproto clients will be seen by MA though.

**Reply by Slim1399** [reply]
I'm using LMS on piCorePlayer right now and MA didn't detect anything !

**Reply by OzGav** [reply]
https://www.music-assistant.io/faq/troubleshooting/ I use picoreplayers every day. If you literally mean you have the LMS option turned on then turn that off.

**Reply by Slim1399** [reply]
When I turned it off I see it, but I couldn't see it on both servers at the same time ?

**Reply by OzGav** [reply]
You shouldn’t be trying to run LMS and MA at the same time. Pick one or the other.

**Reply by Slim1399** [reply]
If I try to do it it's just to integrate the Squeezelite X into the multiroom system, otherwise I have no reason to do it

**Reply by Slim1399** [reply]
And i don't knowed how to install simple client !

**Reply by OzGav** [reply]
Read the docs https://www.music-assistant.io/faq/stream-to/

**Reply by schmurtzm** [reply]
@Slim1399 Did you solve your problems ?  Are you still using Squeezelite-X (I'm a user of it too) ? Di you kept your LMS server or the slimproto from Music Assistant ?

**Comment by Wakkler** [comment]
Is support for the physical buttons of Squeezebox Players (aka Squeezebox Radio) planned? This would be really cool and would allow me to finally switch from LMS to Music Assistant.

**Reply by OzGav** [reply]
Not by the core team https://www.music-assistant.io/player-support/squeezelite/

**Comment by rommess** [comment]
Hi,
I'm using Lyrion Media Server at http://lms.local:9000/ (with 1 queezebox radio and 1 pi core player)  and MA at http://homeassistant.local:8123/d5369777_music_assistant All belong to the same subnet with no restriction.

In HA, i do not have the squeezebox integration.
In MA, i have Chromecast player showing up but nothing in Squeezelite (default settings)

Restart or reinstall did not help.
Am i missing something ?
`2026-02-18 19:57:00.505 INFO (MainThread) [music_assistant] Loaded player provider Squeezelite (slimproto players)
2026-02-18 19:57:28.532 INFO (MainThread) [music_assistant] Loaded player provider Squeezelite (slimproto players)
2026-02-18 20:00:19.259 ERROR (ThreadPoolExecutor-0_4) [music_assistant] Error doing task: Task was destroyed but it is pending!
2026-02-18 20:00:19.259 ERROR (ThreadPoolExecutor-0_4) [music_assistant] Error doing task: Task was destroyed but it is pending!`

**Reply by OzGav** [reply]
Certainly Picoreplayer does not like working with more than one server. The squeezebox might be the same.

**Reply by rommess** [reply]
you're right, I thought that he act like an integration.
so stopping LMS, restarting clients and voila
thanks

**Comment by Archy54** [comment]
I had a huge problem getting my pc to be a music renderer, so far I've landed on WindowsSpin from github as just working straight away. Would prefer a more mainstream compatible renderer if I wanted to use other devices but so far so good if anyone has troubles with squeezelite. Never thought I'd have so much issue just letting my pc run a service to play music randomly from music assistant. Sorry that this of offtopic but this is a workaround if anyone has major issues like I did as windowsspin worked within a minute. Just wish it had volume control but oh well.

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
